google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/AddPerformanceMaxProductListingGroupTree.java

```diff
@@ -0,0 +1,554 @@
+      // The option to remove the existing listing group tree. If the current AssetGroup already has
+      // a tree of ListingGroupFilters, and you try to add a new set of ListingGroupFilters
+      // including a root filter, you'll receive an
+      // 'ASSET_GROUP_LISTING_GROUP_FILTER_ERROR_MULTIPLE_ROOTS' error. Setting this option to true
+      // will remove the existing tree and prevent this error.
+      params.replaceExistingTree = Boolean.parseBoolean("INSERT_REPLACE_EXISTING_TREE_HERE");
+    }
+
+    GoogleAdsClient googleAdsClient = null;
+    try {
+      googleAdsClient = GoogleAdsClient.newBuilder().fromPropertiesFile().build();
+    } catch (FileNotFoundException fnfe) {
+      System.err.printf(
+          "Failed to load GoogleAdsClient configuration from file. Exception: %s%n", fnfe);
+      System.exit(1);
+    } catch (IOException ioe) {
+      System.err.printf("Failed to create GoogleAdsClient. Exception: %s%n", ioe);
+      System.exit(1);
+    }
+
+    try {
+      new AddPerformanceMaxProductListingGroupTree()
+          .runExample(
+              googleAdsClient, params.customerId, params.assetGroupId, params.replaceExistingTree);
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+      System.exit(1);
+    }
+  }
+
+  /**
+   * A factory that creates MutateOperations for removing an existing tree of
+   * AssetGroupListingGroupFilters.
+   *
+   * <p>AssetGroupListingGroupFilters must be removed in a specific order: all of the children of a
+   * filter must be removed before the filter itself, otherwise the API will return an error.
+   *
+   * <p>This object is intended to be used with an array of MutateOperations to perform a series of
+   * related updates to an AssetGroup.
+   */
+  private static class AssetGroupListingGroupFilterRemoveOperationFactory {
+    private String rootResourceName = "";
+    private final HashMap<String, HashSet<String>> parentsToChildren = new HashMap<>();
+
+    public AssetGroupListingGroupFilterRemoveOperationFactory(
+        List<AssetGroupListingGroupFilter> resources) throws Exception {
+      if (resources.isEmpty()) {
+        throw new Exception("No listing group filters to remove");
+      }
+
+      for (AssetGroupListingGroupFilter filter : resources) {
+        // When the node has no parent, it means it's the root node, which is treated differently.
```

Comment:
For readability, would it be worth just handling the root node outside of this loop via something like this?

```
stream over resources -> filter where parent is empty -> collect in a list rootFilters
if rootFilters.size() != 1 throw IllegalStateException
else set root to sole element of rootFilters
```

You'd still have to skip the root in the loop, but perhaps you could use a stream for the loop part and make this more concise. WDYT? Would that make it harder or easier for readers?

Reply:
Updated version strikes a nice balance, so marking this resolved